### PR TITLE
Add the API for search files

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Add the API for search files

#### Release Note
```release-note
Added API for files search (Behind a feature flag)
```

### CODE REVIEW

**Comment by esethna** [comment]
@ethervoid @mgdelacroix can you guys please help with reviews on this when possible to unblock Jesus on this feature?

File: app/migrations.go [hunk 1]
```diff
@@ -283,6 +284,32 @@ func (a *App) DoSystemConsoleRolesCreationMigration() {
...
+func (a *App) doContentExtractionConfigMigration() {
+	// If the migration is already marked as completed, don't do it again.
+	if _, err := a.Srv().Store.System().GetByName(ContentExtractionConfigMigrationKey); err == nil {
+		return
+	}
+
+	if usersCount, err := a.Srv().Store.User().Count(model.UserCountOptions{}); err != nil {
+		mlog.Critical("Failed to get the users count for migrating the content extraction, using default value", mlog.Err(err))
+	} else {
+		if usersCount < 500 {
```
Comment (ethervoid) [inline app/migrations.go:296]: Is this a hard limit we're setting right? Can we extract this number as a constant?

**Review by ethervoid — COMMENTED** [review]
Some comments

File: api4/file.go [hunk 2]
```diff
@@ -725,3 +727,71 @@ func writeFileResponse(filename string, contentType string, contentSize int64, l
...
+	results, err := c.App.SearchFilesInTeamForUser(terms, c.App.Session().UserId, c.Params.TeamId, isOrSearch, includeDeletedChannels, timeZoneOffset, page, perPage)
+
+	elapsedTime := float64(time.Since(startTime)) / float64(time.Second)
+	metrics := c.App.Metrics()
+	if metrics != nil {
+		metrics.IncrementFilesSearchCounter()
+		metrics.ObserveFilesSearchDuration(elapsedTime)
+	}
+
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jespino) [reply api4/file.go:790]: Strictly based on the pre-existing code located in `api4/post.go` if this is wrong, we should do the changes later and consistently across all the searches.

File: app/file.go [hunk 3]
```diff
@@ -771,6 +772,26 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
 		}
 	}
 
+	if *a.Config().FileSettings.ExtractContent {
+		infoCopy := *t.fileinfo
+		a.Srv().Go(func() {
+			file, aerr := a.FileReader(t.fileinfo.Path)
+			if aerr != nil {
+				mlog.Error("Failed to open file for extract file content", mlog.Err(aerr))
```
[the remark this answers is not in the document]
Comment (jespino) [reply app/file.go:780]: good point.

File: app/migrations.go (hunk 1 above)
Comment (jespino) [reply app/migrations.go:296]: Good point.

File: app/file.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jespino) [reply app/file.go:780]: done

File: app/migrations.go (hunk 1 above)
Comment (jespino) [reply app/migrations.go]: done

File: app/upload.go [hunk 4]
```diff
@@ -295,6 +296,21 @@ func (a *App) UploadData(us *model.UploadSession, rd io.Reader) (*model.FileInfo
...
 	}
 
+	if *a.Config().FileSettings.ExtractContent {
+		infoCopy := *info
+		a.Srv().Go(func() {
+			text, err := docextractor.Extract(infoCopy.Name, file, docextractor.ExtractSettings{
+				ArchiveRecursion: *a.Config().FileSettings.ArchiveRecursion,
+			})
+			if err != nil {
+				mlog.Error("Failed to extract file content", mlog.Err(err))
```
Comment (ethervoid) [inline app/upload.go:306]: I missed to mention this one to include the return

**Review by ethervoid — APPROVED** [review]
Another return more and LGTM! :D

File: app/upload.go (hunk 4 above)
Comment (jespino) [reply app/upload.go:306]: Done

File: api4/file.go [hunk 5]
```diff
@@ -725,3 +727,76 @@ func writeFileResponse(filename string, contentType string, contentSize int64, l
...
+		isOrSearch = *params.IsOrSearch
+	}
+
+	page := 0
+	if params.Page != nil {
+		page = *params.Page
+	}
+
+	perPage := 60
+	if params.PerPage != nil {
```
Comment (mgdelacroix) [inline api4/file.go:775]: I think we should check for a max items per page here, something like:

```suggestion
	if params.PerPage != nil && *params.PerPage < 60 {
```

and if we make this change, I'd suggest to move that `60` to a `const`

**Review by mgdelacroix — APPROVED** [review]
This is going to be great!! Adding just one nonblocking nit

File: api4/file.go (hunk 5 above)
Comment (jespino) [reply api4/file.go:775]: This is strictly based on what is done with posts, If you feel strong about this file a ticket to fix this in all the searches.

**Comment by jespino** [comment]
@ogi-m this shouldn't affect anything without the feature flag, and we can test the whole feature later. I would check uploading files, if that works, I think is pretty safe to say that this is not breaking things..

**Review by ogi-m — APPROVED** [review]
Tested with various file types, file upload working as expected 👍🏻

**Pull request merged**